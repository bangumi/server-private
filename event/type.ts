export const EventOp = Object.freeze({
  Create: 'c',
  Delete: 'd',
  Update: 'u',
  Snapshot: 'r',
} as const);

export type EventOp = (typeof EventOp)[keyof typeof EventOp];

export interface Payload {
  // before: object;
  // after: object;
  source: {
    table: string;
  };
  op: EventOp;
}

export interface KafkaMessage {
  topic: string;
  key: string;
  value: string;
}
