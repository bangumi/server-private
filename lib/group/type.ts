export enum GroupTopicMode {
  All = 'all',
  Joined = 'joined',
}

export enum GroupSort {
  Trends = 'trends',
  Created = 'created',
  Updated = 'updated',
}
