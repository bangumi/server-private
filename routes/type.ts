import type { IncomingMessage, ServerResponse } from 'node:http';

import type { Static, TSchema } from '@sinclair/typebox';
import type {
  ContextConfigDefault,
  FastifyBaseLogger,
  FastifyInstance,
  FastifyReply,
  FastifyRequest,
  FastifySchema,
  FastifyTypeProvider,
  RawReplyDefaultExpression,
  RawRequestDefaultExpression,
  RawServerDefault,
  RouteGenericInterface,
  RouteHandlerMethod,
} from 'fastify';
import type { ResolveFastifyRequestType } from 'fastify/types/type-provider.d.ts';

// from https://github.com/fastify/fastify-type-provider-typebox/blob/v3.1.0/index.ts#L56
export interface TypeBoxTypeProvider extends FastifyTypeProvider {
  output: this['input'] extends TSchema ? Static<this['input']> : never;
}

export type App = FastifyInstance<
  RawServerDefault,
  IncomingMessage,
  ServerResponse,
  FastifyBaseLogger,
  TypeBoxTypeProvider
>;

export interface Option {
  tags?: string[];
}

export type Request<schema extends FastifySchema> = FastifyRequest<
  RouteGenericInterface,
  RawServerDefault,
  RawRequestDefaultExpression,
  schema,
  TypeBoxTypeProvider,
  ContextConfigDefault,
  FastifyBaseLogger,
  ResolveFastifyRequestType<TypeBoxTypeProvider, schema, RouteGenericInterface>
>;

export type Reply<schema extends FastifySchema> = FastifyReply<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGenericInterface,
  ContextConfigDefault,
  schema,
  TypeBoxTypeProvider
>;

export type Handler<schema extends FastifySchema> = RouteHandlerMethod<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGenericInterface,
  ContextConfigDefault,
  schema,
  TypeBoxTypeProvider
>;
