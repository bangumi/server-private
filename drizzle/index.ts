export * from './db.ts';
export * as orm from './orm.ts';
export * as schema from './schema.ts';
export * as op from 'drizzle-orm';
