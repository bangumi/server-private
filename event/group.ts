import { getSlimCacheKey, getTopicCacheKey } from '@app/lib/group/cache.ts';
import redis from '@app/lib/redis.ts';
import { getJoinedGroupsCacheKey } from '@app/lib/user/cache.ts';

import { EventOp, type KafkaMessage } from './type';

interface GroupKey {
  grp_id: number;
}

interface GroupPayload {
  op: EventOp;
}

export async function handle({ key, value }: KafkaMessage) {
  const idx = JSON.parse(key) as GroupKey;
  const payload = JSON.parse(value) as GroupPayload;
  switch (payload.op) {
    case EventOp.Create: {
      break;
    }
    case EventOp.Update:
    case EventOp.Delete: {
      await redis.del(getSlimCacheKey(idx.grp_id));
      break;
    }
    case EventOp.Snapshot: {
      break;
    }
  }
}

interface GroupMemberPayload {
  op: EventOp;
  before?: {
    gmb_uid: number;
    gmb_gid: number;
  };
  after?: {
    gmb_uid: number;
    gmb_gid: number;
  };
}

export async function handleMember({ value }: KafkaMessage) {
  const payload = JSON.parse(value) as GroupMemberPayload;
  const uid = payload.before?.gmb_uid ?? payload.after?.gmb_uid;
  if (!uid) {
    return;
  }

  switch (payload.op) {
    case EventOp.Create:
    case EventOp.Update:
    case EventOp.Delete: {
      await redis.del(getJoinedGroupsCacheKey(uid));
      break;
    }
    case EventOp.Snapshot: {
      break;
    }
  }
}

interface GroupTopicKey {
  grp_tpc_id: number;
}

interface GroupTopicPayload {
  op: EventOp;
}

export async function handleTopic({ key, value }: KafkaMessage) {
  const idx = JSON.parse(key) as GroupTopicKey;
  const payload = JSON.parse(value) as GroupTopicPayload;
  switch (payload.op) {
    case EventOp.Create: {
      break;
    }
    case EventOp.Update:
    case EventOp.Delete: {
      await redis.del(getTopicCacheKey(idx.grp_tpc_id));
      break;
    }
  }
}
