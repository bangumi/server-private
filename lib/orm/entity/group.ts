import { Column, Entity, Index, PrimaryGeneratedColumn } from 'typeorm';

import { BooleanTransformer, htmlEscapedString } from '@app/lib/orm/transformer.ts';

import type { User } from './user';

@Entity('chii_groups', { schema: 'bangumi' })
export class Group {
  @PrimaryGeneratedColumn({ type: 'smallint', name: 'grp_id', unsigned: true })
  id!: number;

  @Column('smallint', { name: 'grp_cat', unsigned: true, default: () => "'0'" })
  cat!: number;

  @Column('char', { name: 'grp_name', length: 50 })
  name!: string;

  @Column('char', { name: 'grp_title', length: 50 })
  title!: string;

  @Column('varchar', { name: 'grp_icon', length: 255 })
  icon!: string;

  @Column('mediumint', {
    name: 'grp_creator',
    unsigned: true,
    default: () => "'0'",
  })
  grpCreator!: number;

  @Column('mediumint', {
    name: 'grp_topics',
    unsigned: true,
    default: () => "'0'",
  })
  grpTopics!: number;

  @Column('mediumint', {
    name: 'grp_posts',
    unsigned: true,
    default: () => "'0'",
  })
  grpPosts!: number;

  @Column('mediumint', {
    name: 'grp_members',
    unsigned: true,
    default: () => "'1'",
  })
  memberCount!: number;

  @Column('text', { name: 'grp_desc' })
  description!: string;

  @Column('int', { name: 'grp_lastpost', unsigned: true })
  lastPost!: number;

  @Column('int', { name: 'grp_builddate', unsigned: true })
  builddate!: number;

  @Column('tinyint', {
    name: 'grp_accessible',
    comment: '可访问性',
    width: 1,
    default: () => "'1'",
  })
  accessible!: boolean;

  @Column('tinyint', {
    name: 'grp_nsfw',
    unsigned: true,
    transformer: BooleanTransformer,
  })
  nsfw!: boolean;
}

@Entity('chii_group_members', { schema: 'bangumi' })
export class GroupMembers {
  @Column('mediumint', { primary: true, name: 'gmb_uid', default: () => "'0'" })
  gmbUid!: number;

  @Column('smallint', { primary: true, name: 'gmb_gid', default: () => "'0'" })
  gmbGid!: number;

  @Column('tinyint', { name: 'gmb_moderator', width: 1, default: () => "'0'" })
  gmbModerator!: boolean;

  @Column('int', { name: 'gmb_dateline', unsigned: true, default: () => "'0'" })
  gmbDateline!: number;
}

@Index('grp_tpc_gid', ['parentID'], {})
@Index('grp_tpc_display', ['display'], {})
@Index('grp_tpc_uid', ['creatorID'], {})
@Index('grp_tpc_lastpost', ['updatedAt'], {})
@Entity('chii_group_topics', { schema: 'bangumi' })
export class GroupTopic {
  @PrimaryGeneratedColumn({
    type: 'mediumint',
    name: 'grp_tpc_id',
    unsigned: true,
  })
  id!: number;

  @Column('mediumint', { name: 'grp_tpc_gid', unsigned: true })
  parentID!: number;

  @Column('mediumint', { name: 'grp_tpc_uid', unsigned: true })
  creatorID!: number;

  @Column('varchar', { name: 'grp_tpc_title', length: 80, transformer: htmlEscapedString })
  title!: string;

  @Column('int', {
    name: 'grp_tpc_dateline',
    unsigned: true,
    default: () => "'0'",
  })
  createdAt!: number;

  @Column('int', {
    name: 'grp_tpc_lastpost',
    unsigned: true,
    default: () => "'0'",
  })
  updatedAt!: number;

  @Column('mediumint', {
    name: 'grp_tpc_replies',
    unsigned: true,
    default: () => "'0'",
  })
  replies!: number;

  @Column('tinyint', { name: 'grp_tpc_state', unsigned: true })
  state!: number;

  @Column('tinyint', {
    name: 'grp_tpc_display',
    unsigned: true,
    default: () => "'1'",
  })
  display!: number;

  creator!: User;
}

@Index('pss_topic_id', ['topicID'], {})
@Index('grp_pst_related', ['related'], {})
@Index('grp_pst_uid', ['uid'], {})
@Entity('chii_group_posts', { schema: 'bangumi' })
export class GroupPost {
  @PrimaryGeneratedColumn({
    type: 'mediumint',
    name: 'grp_pst_id',
    unsigned: true,
  })
  id!: number;

  @Column('mediumint', { name: 'grp_pst_mid', unsigned: true })
  topicID!: number;

  @Column('mediumint', { name: 'grp_pst_uid', unsigned: true })
  uid!: number;

  @Column('mediumint', {
    name: 'grp_pst_related',
    comment: '关联回复ID',
    unsigned: true,
    default: () => "'0'",
  })
  related!: number;

  @Column('mediumtext', { name: 'grp_pst_content' })
  content!: string;

  @Column('tinyint', { name: 'grp_pst_state', unsigned: true })
  state!: number;

  @Column('int', {
    name: 'grp_pst_dateline',
    unsigned: true,
    default: () => "'0'",
  })
  dateline!: number;

  creator!: User;
}
