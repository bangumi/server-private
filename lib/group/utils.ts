import { db, op, schema } from '@app/drizzle';

/** Is user member of group */
export async function isMemberInGroup(userID: number, groupID: number): Promise<boolean> {
  const [d] = await db
    .select({ uid: schema.chiiGroupMembers.uid, gid: schema.chiiGroupMembers.gid })
    .from(schema.chiiGroupMembers)
    .where(
      op.and(
        op.eq(schema.chiiGroupMembers.uid, userID),
        op.eq(schema.chiiGroupMembers.gid, groupID),
      ),
    )
    .limit(1);

  return Boolean(d);
}
