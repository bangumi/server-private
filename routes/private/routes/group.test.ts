import { afterEach, beforeEach, describe, expect, test } from 'vitest';

import { db, op, schema } from '@app/drizzle';
import { emptyAuth } from '@app/lib/auth/index.ts';
import { createTestServer } from '@app/tests/utils.ts';

import { setup } from './group.ts';
import { CommentState } from '@app/lib/topic/type.ts';

describe('group list', () => {
  test('should get group list', async () => {
    const app = await createTestServer();
    await app.register(setup);

    const res = await app.inject({
      url: '/groups',
      query: {
        sort: 'created',
        limit: '2',
      },
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('group info', () => {
  test('should get group info', async () => {
    const app = await createTestServer();
    await app.register(setup);

    const res = await app.inject('/groups/sandbox');
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get group members', async () => {
    const app = await createTestServer();
    await app.register(setup);

    const res = await app.inject('/groups/sandbox/members');
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });
});

describe('group topics', () => {
  const testGroupID = 4215;
  const testTopicID = 100;
  const testTopicPostID = 100;
  const testUserID = 382951;
  const testPostID = 101;

  beforeEach(async () => {
    await db.delete(schema.chiiGroupTopics).where(op.eq(schema.chiiGroupTopics.gid, testGroupID));
    await db.delete(schema.chiiGroupPosts).where(op.eq(schema.chiiGroupPosts.mid, testTopicID));
    await db.insert(schema.chiiGroupTopics).values({
      id: testTopicID,
      gid: testGroupID,
      uid: testUserID,
      title: 'Test Topic',
      createdAt: 1462335911,
      updatedAt: 1462335911,
      replies: 1,
      state: 0,
      display: 1,
    });
    await db.insert(schema.chiiGroupPosts).values({
      id: testTopicPostID,
      mid: testTopicID,
      uid: testUserID,
      content: 'Test Topic Content',
      related: 0,
      state: 0,
      createdAt: 1462335911,
    });
    await db.insert(schema.chiiGroupPosts).values({
      id: testPostID,
      mid: testTopicID,
      uid: testUserID,
      content: 'Test Reply',
      related: 0,
      state: 0,
      createdAt: 1462335911,
    });
  });

  afterEach(async () => {
    await db.delete(schema.chiiGroupTopics).where(op.eq(schema.chiiGroupTopics.gid, testGroupID));
    await db.delete(schema.chiiGroupPosts).where(op.eq(schema.chiiGroupPosts.mid, testTopicID));
  });

  test('should get recent topics of joined group', async () => {
    const app = await createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics',
      query: {
        mode: 'joined',
        limit: '2',
      },
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get recent topics of all group', async () => {
    const app = await createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: 287622,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: '/groups/-/topics',
      query: {
        mode: 'all',
        limit: '2',
      },
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get group topics', async () => {
    const app = await createTestServer();
    await app.register(setup);

    const res = await app.inject({
      url: `/groups/sandbox/topics`,
      query: { limit: '2', offset: '0' },
    });
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should create new topic', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/groups/sandbox/topics`,
      method: 'post',
      payload: {
        title: 'New Topic',
        content: 'New Content',
        turnstileToken: 'fake-response',
      },
    });

    expect(res.statusCode).toBe(200);
    const { id } = res.json() as { id: number };

    const [topic] = await db
      .select()
      .from(schema.chiiGroupTopics)
      .where(op.eq(schema.chiiGroupTopics.id, id));
    expect(topic?.title).toBe('New Topic');

    const [post] = await db
      .select()
      .from(schema.chiiGroupPosts)
      .where(op.eq(schema.chiiGroupPosts.mid, id))
      .limit(1);
    expect(post?.content).toBe('New Content');
  });

  test('should edit own topic', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/groups/-/topics/${testTopicID}`,
      method: 'put',
      payload: {
        title: 'Updated Title',
        content: 'Updated Content',
      },
    });

    expect(res.statusCode).toBe(200);

    const [topic] = await db
      .select()
      .from(schema.chiiGroupTopics)
      .where(op.eq(schema.chiiGroupTopics.id, testTopicID));
    expect(topic?.title).toBe('Updated Title');
  });

  test('should not edit topic by non-owner', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID + 1,
      },
    });
    await app.register(setup);

    const res = await app.inject({
      url: `/groups/-/topics/${testTopicID}`,
      method: 'put',
      payload: {
        title: 'Unauthorized Update',
        content: 'Unauthorized Content',
      },
    });

    expect(res.statusCode).toBe(401);
    expect(res.json()).toMatchSnapshot();
  });

  test('should get group post', async () => {
    const app = await createTestServer();
    await app.register(setup);

    const res = await app.inject(`/groups/-/posts/${testPostID}`);
    expect(res.statusCode).toBe(200);
    expect(res.json()).toMatchSnapshot();
  });

  test('should create/edit/delete new post', async () => {
    const app = createTestServer({
      auth: {
        ...emptyAuth(),
        login: true,
        userID: testUserID,
      },
    });
    await app.register(setup);

    // create post
    const createRes = await app.inject({
      url: `/groups/-/topics/${testTopicID}/replies`,
      method: 'post',
      payload: {
        content: 'New Reply',
        turnstileToken: 'fake-response',
      },
    });
    expect(createRes.statusCode).toBe(200);
    const { id } = createRes.json() as { id: number };
    const [createdPost] = await db
      .select()
      .from(schema.chiiGroupPosts)
      .where(op.eq(schema.chiiGroupPosts.id, id));
    expect(createdPost?.content).toBe('New Reply');
    expect(createdPost?.related).toBe(0);

    // update post
    const updateRes = await app.inject({
      url: `/groups/-/posts/${id}`,
      method: 'put',
      payload: {
        content: 'Updated Reply',
      },
    });
    expect(updateRes.statusCode).toBe(200);
    const [updatedPost] = await db
      .select()
      .from(schema.chiiGroupPosts)
      .where(op.eq(schema.chiiGroupPosts.id, id));
    expect(updatedPost?.content).toBe('Updated Reply');

    // delete post
    const deleteRes = await app.inject({
      url: `/groups/-/posts/${id}`,
      method: 'delete',
    });
    expect(deleteRes.statusCode).toBe(200);
    const [deletedPost] = await db
      .select()
      .from(schema.chiiGroupPosts)
      .where(op.eq(schema.chiiGroupPosts.id, id));
    expect(deletedPost?.state).toBe(CommentState.UserDelete);
  });
});
