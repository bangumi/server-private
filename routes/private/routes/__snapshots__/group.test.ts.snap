// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`group info > should get group info 1`] = `
Object {
  "accessible": true,
  "cat": 32,
  "createdAt": 1531631310,
  "creator": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
      "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
      "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
    },
    "group": 0,
    "id": 287622,
    "joinedAt": 0,
    "nickname": "nickname 287622",
    "sign": "sing 287622",
    "username": "287622",
  },
  "creatorID": 287622,
  "description": "[s]非[/s]官方沙盒",
  "icon": Object {
    "large": "https://lain.bgm.tv/pic/icon/l/000/00/42/4215.jpg?r=1531631345",
    "medium": "https://lain.bgm.tv/pic/icon/m/000/00/42/4215.jpg?r=1531631345",
    "small": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
  },
  "id": 4215,
  "members": 3,
  "name": "sandbox",
  "nsfw": false,
  "posts": 0,
  "title": "沙盒",
  "topics": 1,
}
`;

exports[`group info > should get group members 1`] = `
Object {
  "data": Array [
    Object {
      "joinedAt": 1640499538,
      "moderator": false,
      "uid": 382951,
      "user": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
          "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
          "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
        },
        "group": 10,
        "id": 382951,
        "joinedAt": 1512262276,
        "nickname": "树洞酱",
        "sign": "treeholechan@gmail.com 密码:lovemeplease",
        "username": "382951",
      },
    },
    Object {
      "joinedAt": 1531631570,
      "moderator": false,
      "uid": 427613,
      "user": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/icon.jpg",
          "medium": "https://lain.bgm.tv/pic/user/m/icon.jpg",
          "small": "https://lain.bgm.tv/pic/user/s/icon.jpg",
        },
        "group": 0,
        "id": 427613,
        "joinedAt": 0,
        "nickname": "nickname 427613",
        "sign": "sing 427613",
        "username": "427613",
      },
    },
  ],
  "total": 2,
}
`;

exports[`group list > should get group list 1`] = `
Object {
  "data": Array [
    Object {
      "accessible": true,
      "createdAt": 1531631310,
      "creatorID": 287622,
      "icon": Object {
        "large": "https://lain.bgm.tv/pic/icon/l/000/00/42/4215.jpg?r=1531631345",
        "medium": "https://lain.bgm.tv/pic/icon/m/000/00/42/4215.jpg?r=1531631345",
        "small": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
      },
      "id": 4215,
      "members": 3,
      "name": "sandbox",
      "nsfw": false,
      "title": "沙盒",
    },
    Object {
      "accessible": true,
      "createdAt": 1217542289,
      "creatorID": 64,
      "icon": Object {
        "large": "https://lain.bgm.tv/pic/icon/l/000/00/00/11.jpg",
        "medium": "https://lain.bgm.tv/pic/icon/m/000/00/00/11.jpg",
        "small": "https://lain.bgm.tv/pic/icon/s/000/00/00/11.jpg",
      },
      "id": 11,
      "members": 9452,
      "name": "a",
      "nsfw": false,
      "title": "～技术宅真可怕～",
    },
  ],
  "total": 12,
}
`;

exports[`group topics > should get group post 1`] = `
Object {
  "content": "Test Reply",
  "createdAt": 1462335911,
  "creator": Object {
    "avatar": Object {
      "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
      "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
      "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
    },
    "group": 10,
    "id": 382951,
    "joinedAt": 1512262276,
    "nickname": "树洞酱",
    "sign": "treeholechan@gmail.com 密码:lovemeplease",
    "username": "382951",
  },
  "creatorID": 382951,
  "id": 101,
  "state": 0,
  "topic": Object {
    "createdAt": 1462335911,
    "creator": Object {
      "avatar": Object {
        "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
        "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
        "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
      },
      "group": 10,
      "id": 382951,
      "joinedAt": 1512262276,
      "nickname": "树洞酱",
      "sign": "treeholechan@gmail.com 密码:lovemeplease",
      "username": "382951",
    },
    "creatorID": 382951,
    "display": 1,
    "id": 100,
    "parentID": 4215,
    "replies": 1,
    "state": 0,
    "title": "Test Topic",
    "updatedAt": 1462335911,
  },
}
`;

exports[`group topics > should get group topics 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": 1462335911,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
          "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
          "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
        },
        "group": 10,
        "id": 382951,
        "joinedAt": 1512262276,
        "nickname": "树洞酱",
        "sign": "treeholechan@gmail.com 密码:lovemeplease",
        "username": "382951",
      },
      "creatorID": 382951,
      "display": 1,
      "id": 100,
      "parentID": 4215,
      "replies": 1,
      "state": 0,
      "title": "Test Topic",
      "updatedAt": 1462335911,
    },
  ],
  "total": 1,
}
`;

exports[`group topics > should get recent topics of all group 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": 1462335911,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
          "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
          "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
        },
        "group": 10,
        "id": 382951,
        "joinedAt": 1512262276,
        "nickname": "树洞酱",
        "sign": "treeholechan@gmail.com 密码:lovemeplease",
        "username": "382951",
      },
      "creatorID": 382951,
      "display": 1,
      "group": Object {
        "accessible": true,
        "createdAt": 1531631310,
        "creatorID": 287622,
        "icon": Object {
          "large": "https://lain.bgm.tv/pic/icon/l/000/00/42/4215.jpg?r=1531631345",
          "medium": "https://lain.bgm.tv/pic/icon/m/000/00/42/4215.jpg?r=1531631345",
          "small": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
        },
        "id": 4215,
        "members": 3,
        "name": "sandbox",
        "nsfw": false,
        "title": "沙盒",
      },
      "id": 100,
      "parentID": 4215,
      "replies": Array [],
      "state": 0,
      "title": "Test Topic",
      "updatedAt": 1462335911,
    },
  ],
  "total": 1000,
}
`;

exports[`group topics > should get recent topics of joined group 1`] = `
Object {
  "data": Array [
    Object {
      "createdAt": 1462335911,
      "creator": Object {
        "avatar": Object {
          "large": "https://lain.bgm.tv/pic/user/l/000/38/29/382951.jpg?r=1571167246",
          "medium": "https://lain.bgm.tv/pic/user/m/000/38/29/382951.jpg?r=1571167246",
          "small": "https://lain.bgm.tv/pic/user/s/000/38/29/382951.jpg?r=1571167246",
        },
        "group": 10,
        "id": 382951,
        "joinedAt": 1512262276,
        "nickname": "树洞酱",
        "sign": "treeholechan@gmail.com 密码:lovemeplease",
        "username": "382951",
      },
      "creatorID": 382951,
      "display": 1,
      "group": Object {
        "accessible": true,
        "createdAt": 1531631310,
        "creatorID": 287622,
        "icon": Object {
          "large": "https://lain.bgm.tv/pic/icon/l/000/00/42/4215.jpg?r=1531631345",
          "medium": "https://lain.bgm.tv/pic/icon/m/000/00/42/4215.jpg?r=1531631345",
          "small": "https://lain.bgm.tv/pic/icon/s/000/00/42/4215.jpg?r=1531631345",
        },
        "id": 4215,
        "members": 3,
        "name": "sandbox",
        "nsfw": false,
        "title": "沙盒",
      },
      "id": 100,
      "parentID": 4215,
      "replies": Array [],
      "state": 0,
      "title": "Test Topic",
      "updatedAt": 1462335911,
    },
  ],
  "total": 1000,
}
`;

exports[`group topics > should not edit topic by non-owner 1`] = `
Object {
  "code": "NOT_ALLOWED",
  "error": "Unauthorized",
  "message": "you don't have permission to edit this topic",
  "statusCode": 401,
}
`;
